OpenSim/Simulation/Model/Geometry.cpp

```diff
@@ -171,50 +171,63 @@ void FrameGeometry::implementCreateDecorativeGeometry(SimTK::Array_<SimTK::Decor
-    Array_<string> attempts;
-    const Model& model = getFrame().getModel();
-    bool foundIt = ModelVisualizer::findGeometryFile(model, file, isAbsolutePath, attempts);
-
-    if (!foundIt) {
-        std::clog << "ModelVisualizer couldn't find file '" << file
-            << "'; tried\n";
-        for (unsigned i = 0; i < attempts.size(); ++i)
-            std::clog << "  " << attempts[i] << "\n";
-        if (!isAbsolutePath &&
-            !Pathname::environmentVariableExists("OPENSIM_HOME"))
-            std::clog << "Set environment variable OPENSIM_HOME "
-            << "to search $OPENSIM_HOME/Geometry.\n";
-        return;
-    }
-
-    SimTK::PolygonalMesh pmesh;
-    try {
+    // Current interface to Visualizer calls this method on every frame.
+    // On first time through, load file and create DecorativeMeshFile else
+    // try to reuse it.
+    if (!triedToLoad) {
+        triedToLoad = true;
+        const std::string& file = get_mesh_file();
+        bool isAbsolutePath; string directory, fileName, extension;
+        SimTK::Pathname::deconstructPathname(file,
+            isAbsolutePath, directory, fileName, extension);
+        const string lowerExtension = SimTK::String::toLower(extension);
+        if (lowerExtension != ".vtp" && lowerExtension != ".obj" && lowerExtension != ".stl") {
+            std::clog << "ModelVisualizer ignoring '" << file
+                << "'; only .vtp .stl and .obj files currently supported.\n";
+            return;
+        }
+
+        // File is a .vtp or .obj. See if we can find it.
+        Array_<string> attempts;
+        const Model& model = getFrame().getModel();
+        bool foundIt = ModelVisualizer::findGeometryFile(model, file, isAbsolutePath, attempts);
+
+        if (!foundIt) {
+            std::clog << "ModelVisualizer couldn't find file '" << file
+                << "'; tried\n";
+            for (unsigned i = 0; i < attempts.size(); ++i)
+                std::clog << "  " << attempts[i] << "\n";
+            if (!isAbsolutePath &&
+                !Pathname::environmentVariableExists("OPENSIM_HOME"))
+                std::clog << "Set environment variable OPENSIM_HOME "
+                << "to search $OPENSIM_HOME/Geometry.\n";
+            return;
+        }
+
+        SimTK::PolygonalMesh pmesh;
+        try {
             std::ifstream objFile;
             objFile.open(attempts.back().c_str());
             pmesh.loadFile(attempts.back().c_str());
             // objFile closes when destructed
-        
+
+        }
+        catch (const std::exception& e) {
+            std::clog << "Visualizer couldn't read "
+                << attempts.back() << " because:\n"
+                << e.what() << "\n";
+            return;
+        }
+
+        DecorativeMeshFile* dmesh = new DecorativeMeshFile(attempts.back().c_str());
+        dmesh->setScaleFactors(get_scale_factors());
+        decoGeoms.push_back(*dmesh);
+        cachedMesh.reset(dmesh);
```

Comment:
Lines 225 and 223 should not be repeated; they should just be outside of the `if (!triedToLoad)`:

```
if (!cachedMesh) {
    cachedMesh.reset(new ...);
}
cachedMesh->setScaleFactors(...);
decoGems.push_back(*cachedMesh);
```

I know this all might change anyway (if you move the loading to finalizeFromProperties()).